Permissionless Access node setup doc (#1256)
onflow/flow: docs/content/node-operation/access-node-setup.mdx

---
title: Setting Up a Flow Access Node
sidebar_title: Access Node Setup
description: How to run a permissionless Access node
---

This guide is for running a permissonless Access node on Flow. If you are planning to run a different type of staked node then see [node bootstrap](node-bootstrap,mdx).

Permissionless Access node allow any operator to run a Flow Access node.
Since this is the first time the network is onboarding access nodes via this new process, the number of slots available to run an access node are limited to five.

> Once the first cohort of five access nodes onboard and are able to run successfully, more slots will be made available every epoch.

Author: yeah - rewording this part.

> Once the first cohort of five access nodes onboard and are able to run successfully, more slots will be made available every epoch.

Author: updated text with this info

To run an access node you will need to provision a machine or virtual machine to run your node software. Please see follow the [node-provisioning](node-provisioning.mdx) guide for it.
You can provision the machine before or after your node has been chosen.

At a high level, to run a permissionless Access node, you will have to do the following steps:
1. Generate the node identity (private and public keys, node ID etc.).

> 2. Stake the node with 100 FLOW by EOD Tuesday (date TBD) by providing the node information generated in step 1.

Reviewer: the staking auction usually ends midday tuesday, right? not EOD?

3. You can verify if your node ID was randomly selected by the on-chain selection process on the following Wednesday (see below).

> 4. If your node ID was selected, you can provision and start running the node. If your node wasn't selected, you can leave the tokens staked and try again in the next epoch. Your staked tokens will be returned in that case.

Reviewer: ```suggestion
4. If your node ID was selected, you can provision and start running the node. If your node wasn't selected, your tokens will have been refunded to your unstaked bucket in the staking smart contract. When the next epoch begins, you can try committing tokens again in a future epoch to get a new spot.
```

$tree ./bootstrap/
./bootstrap/
├── private-root-information
│        └── private-node-info_e737ec6efbd26ef43bf676911cdc5a11ba15fc6562d05413e6589fccdd6c06d5
│        ├── node-info.priv.json
│        └── secretsdb-key
└── public-root-information
    ├── node-id
    └── node-info.pub.e737ec6efbd26ef43bf676911cdc5a11ba15fc6562d05413e6589fccdd6c06d5.json

3 directories, 4 files
```

Your node identity has now been generated. Your <b>node ID</b> can be found in the file `./bootstrap/public-root-information/node-id`.

```shell:title=Example
$cat ./bootstrap/public-root-information/node-id
e737ec6efbd26ef43bf676911cdc5a11ba15fc6562d05413e6589fccdd6c06d5
```

<Callout type="warning">
    All your private keys should be in the `bootstrap` folder created earlier. Please take a back up of the entire folder.
</Callout>

## Step  2 - Stake the node

You need to now register the node on chain by staking the node via [Flow Port](https://port.onflow.org/). [Here](https://developers.flow.com/nodes/flow-port) is a guide on how to use Flow port if you are not familiar with it.

> If you are staking via a custody provider or would like to directly submit a staking transaction instead follow this [guide](https://developers.flow.com/nodes/staking#how-do-i-stake).

Reviewer 2: Can we use relative links here, so the links would work on eg. local dev environments?

{
  "Role": "consensus",
  "Address": "consensus-001.nodes.flow.com:3569",
  "NodeID": "39fa54984b8eaa463e129919464f61c8cec3a4389478df79c44eb9bfbf30799a",
  "Weight": 1000,
  "NetworkPubKey": "d92e3d5880abe233cf9fe9104db34bbb31251468a541454722b3870c04156a1b0504aef443bcaad124b997384b8fe7052847ce1e6189af1392d865e6be69835b",
  "StakingPubKey": "917826e018f056a00b778a58ae83054906957ffd4b6f1b7da083551f7a9f35e02b76ace50424ed7d2c9fc69207a59f0f08a031048f5641db94e77d0648b24d150dedd54bab7cd44b4aa60cfd54be418647b0b3965f8ae54c0bcb48ae9d705162"
}
```

### Timing

New nodes are able to join the network each time a new epoch begins.
An epoch is a period of time (approximately one week) when the node operators in the network are constant.
At epoch boundaries, newly staked node operators are able to join the network and existing node operators which have unstaked may exit the network.
You can read more about epochs [here](https://developers.flow.com/nodes/staking/schedule).

In order to join the network at epoch N+1, the node **must** be registered with at least 100 FLOW staked prior to the end of epoch N's Staking Auction Phase.

> Currently, an epoch transition on mainnet starts on Wednesday early morning. Hence, the node should be staked by EOD Tuesday to be able to join the network starting on Wednesday.

Author: @jordanschalm and @joshuahannan - I have explicitly mentioned the day of the week instead of epoch phases here since I wanted to make it easier for new node operators to consume this content. Lemme know what you think.

Author: yup! will add that

Confirmation of a new node's inclusion in epoch N+1 is included in the [`EpochSetup` event](/staking/epoch-scripts-events/#epochsetup).

![Flow Epoch Schedule](https://storage.googleapis.com/flow-resources/documentation-assets/epoch-startup-order.png)

> ## Step 3 - Check if the node ID got selected by the random on-chain selection process

Author: @jordanschalm and @joshuahannan - here is where I need your help. Can you point me to the scripts that the operator should run? Also, how to check the service event?

> ## Step 3 - Check if the node ID got selected by the random on-chain selection process

Reviewer 3: ```suggestion
## Step 3 - Verify if your node ID was selected
```
Simplifying (reduce repetition) and make consistent with above grammar tweaks

> How to check? What scripts to run?

Reviewer: ```suggestion
There are two main ways to verify if your node was selected. If you can monitor events, look for [the epoch setup event](/staking/epoch-scripts-events/#epochsetup) that gets emitted by the epoch contract. That event is emitted at the end of epoch N's staking auction and contains a list of node IDs that are confirmed for the next epoch.

Another way to verify is the query the information directly. After the staking auction ends, you can use the [get node info script](/staking/epoch-scripts-events#get-all-of-the-info-associated-with-a-single-node-staker) to get the information about your node. If your tokens are shown in the `tokensCommitted` or `tokensStaked` balance, you are confirmed for the next epoch. If the tokens are in the `tokensUnstaked` balance, then your node was not chosen.
```

Reviewer 2: Could we write a script specifically for this use-case which provides a more concise answer? For example, accept the node ID as input, read the `proposedEpochMetadata`, and return true if the node ID is in the identity table?

Author: that will be very helpful!

You can read more about epochs [here](https://developers.flow.com/nodes/staking/schedule).

In order to join the network at epoch N+1, the node **must** be registered with at least 100 FLOW staked prior to the end of epoch N's Staking Auction Phase.
Currently, an epoch transition on mainnet starts on Wednesday (Pacific time) early morning. Hence, the node should be staked by EOD Tuesday (Pacific time) to be able to join the network starting on Wednesday.

Confirmation of a new node's inclusion in epoch N+1 is included in the [`EpochSetup` event](/staking/epoch-scripts-events/#epochsetup).

![Flow Epoch Schedule](https://storage.googleapis.com/flow-resources/documentation-assets/epoch-startup-order.png)

## Step 3 - Verify if your node ID was selected

There are two main ways to verify if your node was selected. If you can monitor events, look for [the epoch setup event](/staking/epoch-scripts-events/#epochsetup) that gets emitted by the epoch contract. That event is emitted at the end of epoch N's staking auction and contains a list of node IDs that are confirmed for the next epoch.

> Another way to verify is the query the information directly. After the staking auction ends, you can use the [get node info script](/staking/epoch-scripts-events#get-all-of-the-info-associated-with-a-single-node-staker) to get the information about your node. If your tokens are shown in the `tokensCommitted` or `tokensStaked` balance, you are confirmed for the next epoch. If the tokens are in the `tokensUnstaked` balance, then your node was not chosen.

Author: @jordanschalm and @joshuahannan - can we use the Flow CLI here instead and say something like this?

Check your current staking-info. It it says token Committed then your node has been selected.

> 
> $ flow accounts staking-info 0x32a1b5d0a0bc7f33  -n testnet
> 
> Account staking info:
> 	ID: 						 "5f6c73a22445d7d958c6a37c1f3be99c72cacd39894a3e46d6647a9adb007b4d"
> 	Initial Weight: 				 100
> 	Networking Address: 				 "execution-001.devnet41.nodes.onflow.org:3569"
> 	Networking Key: 				 "5db0d4a2aa1f72d6bd03decfddb3a45199a8f42321dce09a176aed85f8c76e82a969a721968e99dca3204a311b7d681a12b5e97450dfb15ba9f8bfabbd91ce80"
> 	Role: 						 3
> 	Staking Key: 					 "8c30fdf402de3861d2c78189e16e46a9d79ac6eebdfc0c54a92e283999997fb568c65843127dbec0c8b80d153d3167d616b1590844a4a6d678b40335938ded790778a868809f215f521256a422021ab411616ccba03c5a6dfcdf0beb35cc114d"
> 	Tokens Committed: 				 0.00000000
> 	Tokens To Unstake: 				 0.00000000
> 	Tokens Rewarded: 				 7413960.66250000
> 	Tokens Staked: 					 1250000.00000000
> 	Tokens Unstaked: 				 0.00000000
> 	Tokens Unstaking: 				 0.00000000
> 	Node Total Stake (including delegators): 	 1250000.00000000
> 
> Account has no delegations.

Author: I can add examples for before and after, but this should be a way for operators to check if they got selected or not, correct?

Author: updated the doc with cli and flowport example on how to check if node got selected.

## Limitations
This is a preliminary implementation of full permissionless node operation and there will be exactly **five** slots in total that will be opened for access nodes.
If there are more than five staked candidate access nodes at the end of the staking phase on 14th Feb 2023 EOD PT, five of those will be selected by an on-chain random selection process and no other permissionless ANs can join until others unstake in a future epoch.

More slots will be opened in the future if this goes well.

Support for unlimited access nodes requires staking auctions to be implemented, which is still an upgrade planned for the future.

## How to run a Permissionless Access node?

> Note: To run an access node you will need to provision a machine or virtual machine to run your node software. Please follow the [node-provisioning](node-provisioning.mdx) guide for it.
You can provision the machine before or after your node has been chosen.

At a high level, to run a permissionless Access node, you will have to do the following steps:
1. Generate the node identity (private and public keys, node ID etc.).

> 2. Stake the node with 100 FLOW by **EOD Tuesday, 14th February 2023 Pacific time (02/14/2023 11:59 PM or 02/15/2023 7:59 AM UTC)** (see [timing](#timing)) by providing the node information generated in step 1.

Reviewer 2: I feel like the Feb 14 date should go out in an announcement, whereas the documentation should be more general and say something like "each Tuesday EOD". Otherwise it will be outdated almost immediately. 

We could say here that the _first_ slots will open Feb 14.

By the way, do we provide a way for people to look-up whether there are any slots available? After the initial few allocations of slots where we announce them explicitly, a look-up tool/script would be the way people check for available slots.

Author: make sense - lemme reword it

Author: ok - i made the guide more generic but called out the exact time inside the limitations section.

3 directories, 4 files
```

> ⚠️ _Use a fully qualified domain name for the network address. Please also include the port number in the network address e.g. `flowaccess.mycompany.com:3569`_

> ⚠️ _Do not include in `http://` in the network address._

> If you would like to stake multiple access nodes, please ensure you generate a unique identity for each node.

Your node identity has now been generated. Your <b>node ID</b> can be found in the file `./bootstrap/public-root-information/node-id`.

```shell:title=Example
$cat ./bootstrap/public-root-information/node-id
e737ec6efbd26ef43bf676911cdc5a11ba15fc6562d05413e6589fccdd6c06d5
```

> All your private keys should be in the `bootstrap` folder created earlier. Please take a back up of the entire folder.

## Step  2 - Stake the node

You need to now register the node on chain by staking the node via [Flow Port](https://port.onflow.org/).

[Here](https://developers.flow.com/nodes/flow-port/staking-guide) is a guide on how to use Flow port if you are not familiar with it.
If you are staking via a custody provider or would like to directly submit a staking transaction instead follow this [guide](/nodes/staking#how-do-i-stake).

> Fund you Flow account with at least 100.01 FLOW tokens, which covers the required stake plus the storage deposit.

Reviewer: This can be 100.00. it is `>=` in the contract

Author: when testing on Flow port, I funded the account with 100FLOW and the transaction failed with an insufficient funds error since it needed to keep some FLOW in the account for storage.

Reviewer: oh, I read that wrong. I thought you meant register the node. We're all good. :)

You can read more about epochs [here](https://developers.flow.com/nodes/staking/schedule).

In order to join the network at epoch N+1, the node **must** be registered with at least 100 FLOW staked prior to the end of epoch N's Staking Auction Phase.
Currently, an epoch transition on mainnet starts on Wednesday (Pacific time) early morning. Hence, the node should be staked by EOD Tuesday (Pacific time) to be able to join the network starting on Wednesday.

Confirmation of a new node's inclusion in epoch N+1 is included in the [`EpochSetup` event](/staking/epoch-scripts-events/#epochsetup).

![Flow Epoch Schedule](https://storage.googleapis.com/flow-resources/documentation-assets/epoch-startup-order.png)

## Step 3 - Verify that your node ID was selected

On Wednesday, 15th Feb 2023 at ~3:30 AM Pacific (11:30 AM UTC), the staking auction for the current epoch will end and five nodes from candidate list of nodes will be chosen at random by the staking contract to be part of the next epoch.

There are several ways to verify whether your node was chosen as explained below.

> When you stake the node, the tokens will show up under the `tokensCommitted` bucket. After the staking auction ends, if the node is selected, the tokens are moved to the `tokensStaked` bucket.

Reviewer: ```suggestion
When you stake the node, the tokens will show up under the `tokensCommitted` bucket. After the staking auction ends, if the node is selected, the tokens remain in the `tokensCommitted` bucket and are moved to the `tokensStaked` bucket at the end of the epoch.
```

## Limitations
This is a preliminary implementation of full permissionless node operation and there will be exactly **five** slots in total that will be opened for access nodes.
If there are more than five staked access nodes at the end of the staking phase on 14th Feb 2023 EOD PT, five of those will be selected by an on-chain random selection process and no other permissionless ANs can join until others unstake in the next epoch.

More slots will be opened in the future if this goes well.

Support for unlimited access nodes requires staking auctions to be implemented, which is still an upgrade planned for the future.

## How to run a Permissionless Access node?

> Note: To run an access node you will need to provision a machine or virtual machine to run your node software. Please follow the [node-provisioning](node-provisioning.mdx) guide for it.
You can provision the machine before or after your node has been chosen.

At a high level, to run a permissionless Access node, you will have to do the following steps:
1. Generate the node identity (private and public keys, node ID etc.).

> 2. Stake the node with 100 FLOW by **EOD Tuesday, 14th February 2023 Pacific time (02/14/2023 12:00 PM Pacific or 02/15/2023 8:00 AM UTC)** by providing the node information generated in step 1.

Reviewer 4: Is this supposed to be `12:00 AM`?

Reviewer 4: Maybe add a note to see the `Timing` section for details

Author: just changed it to 23:50 to avoid confusion

3 directories, 4 files
```

> ⚠️ _Use a fully qualified domain name for the network address. Please also include the port number in the network address e.g. `flowaccess.mycompany.com:3569`_

> ⚠️ _Do not include in `http://` in the network address._

> If you would like to stake multiple access nodes, please ensure you generate a unique identity for each node.

Your node identity has now been generated. Your <b>node ID</b> can be found in the file `./bootstrap/public-root-information/node-id`.

```shell:title=Example
$cat ./bootstrap/public-root-information/node-id
e737ec6efbd26ef43bf676911cdc5a11ba15fc6562d05413e6589fccdd6c06d5
```

> All your private keys should be in the `bootstrap` folder created earlier. Please take a back up of the entire folder.

## Step  2 - Stake the node

You need to now register the node on chain by staking the node via [Flow Port](https://port.onflow.org/).

[Here](https://developers.flow.com/nodes/flow-port/staking-guide) is a guide on how to use Flow port if you are not familiar with it.
If you are staking via a custody provider or would like to directly submit a staking transaction instead follow this [guide](/nodes/staking#how-do-i-stake).

> Fund you Flow account with at least 100.01 FLOW token required the stake the node and pay for storage.

Reviewer 4: ```suggestion
Fund you Flow account with at least 100.01 FLOW tokens, which covers the required stake plus the storage deposit.
```

> 🚀 The access node should now be up and running, and you should be able to query the node using Flow CLI or curl,

```shell:title=Example
flow blocks get latest --host localhost:9000
```

```shell:title=Example
curl http://localhost/v1/blocks?height=sealed
```

## Monitoring and Metrics

The node publishes several Prometheus metrics. See [Monitoring Node Health](monitoring-nodes) to setup node monitoring.

### Node Status

The metrics for the node should be able to provide a good overview of the status of the node. If we want to get a quick snapshot of the status of the node, and if it's properly participating in the network, you can check the `consensus_compliance_finalized_height` or `consensus_compliance_sealed_height` metric, and ensure that it is not zero and strictly increasing.

```shell
curl localhost:8080/metrics | grep consensus_compliance_sealed_height

# HELP consensus_compliance_sealed_height the last sealed height
# TYPE consensus_compliance_sealed_height gauge
consensus_compliance_sealed_height 1.132054e+06
```

## FAQs

### Will the access node receive rewards?

> No, the access node do not receive any rewards.

Reviewer 4: ```suggestion
No, the access nodes do not receive any rewards.
```

## Timing

New nodes are able to join the network each time a new epoch begins.
An epoch is a period of time (approximately one week) when the node operators in the network are constant.
At epoch boundaries, newly staked node operators are able to join the network and existing node operators which have unstaked may exit the network.
You can read more about epochs [here](https://developers.flow.com/nodes/staking/schedule).

In order to join the network at epoch N+1, the access node **must** be registered with at least 100 FLOW staked prior to the end of epoch N's Staking Auction Phase.

Currently on mainnet, the staking auction starts every Wednesday at around 17:30 UTC and ends on the next Wednesday at 8:30 UTC.
Hence, the node should be staked by _Wednesday, 08:30 UTC_ to be able to join the network in the next epoch.

Confirmation of a new node's inclusion in epoch N+1 is included in the [`EpochSetup` event](/staking/epoch-scripts-events/#epochsetup).

![Flow Epoch Schedule](https://storage.googleapis.com/flow-resources/documentation-assets/epoch-startup-order.png)

> ## Limitations

Author: I will be revising this section after 15th Feb based on what happens. If all slots are taken then I will indicate that.

Author: done

### Can the Access node be unstaked?

Yes, like any other staked node, the Access node can be unstaked. The staked tokens will be moved to the unstaked bucket in the subsequent epoch.

### How to see all the access nodes that have staked?

When the nodes are initially staked, they are all added to the candidate list of nodes before the end of the epoch staking phase.
The list can be retrieved from the chain by executing the [get_candidate_nodes](https://github.com/onflow/flow-core-contracts/blob/48ba17d3386023d70817197a20effbc5d16339b3/transactions/idTableStaking/scripts/get_candidate_nodes.cdc) script which returns the candidate list for the current epoch.

```shell
$ flow scripts execute  ./transactions/idTableStaking/scripts/get_candidate_nodes.cdc -n mainnet
```

### How to check the availability of open access nodes slots for the next epoch?

The limits for the open slots are defined in the staking contract and can be queried from the chain by executing the [get_candidate_limit](https://github.com/onflow/flow-core-contracts/blob/master/transactions/idTableStaking/scripts/get_candidate_limits.cdc) script.

```shell

> $ flow scripts execute ./transactions/idTableStaking/scripts/get_candidate_limits.cdc -n mainnet

Author: @jordanschalm added the script here to find the candidate limits.

## Who should run a Permissionless Access node?
dApp developers can choose to run their own private permissionless access node and move away from using the community access nodes. This will also allow them to not be subjected to the API rate limits of the public access nodes.

Node operators can also run their own permissionless access node and provide access to that node as a service.

Chain analytics, audit and exploration applications can run such an access node and do not have to rely on third parties for the state of the network.

## Timing

New nodes are able to join the network each time a new epoch begins.
An epoch is a period of time (approximately one week) when the node operators in the network are constant.
At epoch boundaries, newly staked node operators are able to join the network and existing node operators which have unstaked may exit the network.
You can read more about epochs [here](https://developers.flow.com/nodes/staking/schedule).

In order to join the network at epoch N+1, the access node **must** be registered with at least 100 FLOW staked prior to the end of epoch N's Staking Auction Phase.

> Currently on mainnet, the staking auction starts every Wednesday at around 17:30 UTC and ends on the next Wednesday at 8:30 UTC.
> Hence, the node should be staked by _Wednesday, 08:30 UTC_ to be able to join the network in the next epoch.

Reviewer 2: ```suggestion
Currently on mainnet, the staking auction starts every Wednesday at around 20:00 UTC and ends on the next Wednesday at around 12:00 UTC. 
Since this deadline may shift slightly from epoch to epoch, we recommend the node be staked by _Wednesday, 8:00 UTC_ to be able to join the network in the next epoch.
```

- 12:00 PT is 20:00UTC when I run the conversion. 
- The target time for staking auction end is 4AM PT or 12:00UTC. I'm OK to either pad this deadline like you have (say 8:00UTC) or to give the real deadline and add a note that the deadlines shift epoch to epoch (added this as suggestion)

Author: @jordanschalm am confused. On grafana, I see different time in UTC - 
![image](https://user-images.githubusercontent.com/1117327/217352546-3a519062-c93c-4a6d-8f7a-b0f530a25f41.png)

Reviewer 2: That graph is showing the actual phase change time, for one epoch (the last one). Last week the transition was earlier than the target -- over the course of the last week we've adjusted, and we're tracking for about a 11:30UTC staking auction end (target is 12:00UTC).

Confirmation of a new node's inclusion in epoch N+1 is included in the [`EpochSetup` event](/staking/epoch-scripts-events/#epochsetup).

![Flow Epoch Schedule](https://storage.googleapis.com/flow-resources/documentation-assets/epoch-startup-order.png)

## Limitations
This is a preliminary implementation of the full permissionless node operation feature and there will be exactly **five** slots in total that will be opened for access nodes.

> If there are more than five staked candidate access nodes at the end of the staking phase of the current epoch which ends on Wedensday, 15th Feb, five of those will be selected by an on-chain random selection process and no other permissionless ANs can join until others unstake in a future epoch.

Reviewer 2: ```suggestion
If there are more than five staked candidate access nodes at the end of the staking phase of the current epoch which ends on Wednesday, 15th Feb, five of those will be selected by an on-chain random selection process and no other permissionless ANs can join until others unstake in a future epoch.
```

> More slots will be opened in the future if this goes well.

Reviewer 2: ```suggestion
More slots will be opened in the future.
```
😅

Support for unlimited access nodes requires staking auctions to be implemented, which is still an upgrade planned for the future.

To summarize,

|  **Date**  |  **Time** | **Epoch** |     **Epoch Phase**    |                                                     |
|:----------:|:---------:|:---------:|:----------------------:|:---------------------------------------------------:|

> | 02/08/2023 | 17:30 UTC | n         | Staking auction starts | Stake your access nodes                             |
> | 02/15/2023 | 08:00 UTC | n         | Staking auction ends   | No more access nodes can be staked for epoch n+1    |
> | 02/15/2023 | 17:30 UTC | n+1       | Epoch n+1 starts       | 5 randomly chosen access nodes can join the network |

Reviewer 2: ```suggestion
| 02/08/2023 | 20:00 UTC | 63         | Staking auction starts | Stake your access nodes                             |
| 02/15/2023 | 08:00 UTC | 63         | Staking auction ends   | No more access nodes can be staked for epoch 64    |
| 02/15/2023 | 20:00 UTC | 64       | Epoch n+1 starts       | 5 randomly chosen access nodes can join the network |
```

> > Please ensure that you have staked the access node by **15th Feb, 2023 Wednesday 8:00 AM UTC** if you wish to run the node in the next epoch.

Reviewer 2: ```suggestion
> Please ensure that you have staked the access node by **15th Feb, 2023 Wednesday 08:00 UTC** if you wish to run the node in the next epoch.
```

## How to run a Permissionless Access node?

> Note: To run an access node you will need to provision a machine or virtual machine to run your node software. Please follow the [node-provisioning](node-provisioning.mdx) guide for it.
You can provision the machine before or after your node has been chosen.

At a high level, to run a permissionless Access node, you will have to do the following steps:
1. Generate the node identity (private and public keys, node ID etc.).
2. Stake the node with 100 FLOW by the end of the staking phase of the current epoch (see [timing](#timing)) by providing the node information generated in step 1.

> 3. You can verify if your node ID was selected by the on-chain random selection process on Wednesday at around 17:30 UTC when the next epoch starts.

Reviewer 2: ```suggestion
3. You can verify if your node ID was selected by the on-chain random selection process on Wednesday at around 20:00 UTC when the next epoch starts.
```

Those node details (`Node ID`, `Network Address`, `Networking Key` and `Staking Key`) can be found in the file: `./bootstrap/public-root-information/node-info.pub.<node-id>.json`.

```shell:title=Example
$cat ./bootstrap/public-root-information/node-info.pub. e737ec6efbd26ef43bf676911cdc5a11ba15fc6562d05413e6589fccdd6c06d5.json
{
  "Role": "access",
  "Address": "flowaccess.mycompany.com:3569",
  "NodeID": "e737ec6efbd26ef43bf676911cdc5a11ba15fc6562d05413e6589fccdd6c06d5",
  "Weight": 0,
  "NetworkPubKey": "f493a74704f6961ae7903e062ecd58d990672858eff99aece7bfbccf3aa02de8f1a624ecbf21a01e8b2f4a5854c231fbe218edd7762a34fea881f3958a215305",
  "StakingPubKey": "ae8dcf81f3a70d72036b7ba2c586ed37ed0eb82b9c0a4aab998a8420f98894f94c14f84fa716e93654d3940fc0c8ff4d19b504c90a5b4918b28f421e9d3659dc2b7e246025ebeffea0d83cceefe315d7ed346dbe412fdac51b64997d97d29f7e"
}
```

#### Example

![node_details_permissionless_an](node_details_permissionless_an.png)

On the next screen, ensure that you stake 100 FLOW token.

#### Example

![transaction_register_node_permissionless_an](transaction_register_node_permissionless_an.png)

Submit the Transaction.

## Step 3 - Verify that your node ID was selected

> On Wednesday at 17:30 AM UTC, the staking auction for the current epoch will end and five nodes from candidate list of nodes will be chosen at random by the staking contract to be part of the next epoch.

Reviewer 2: ```suggestion
On Wednesday at around 12:00 UTC, the staking auction for the current epoch will end and five nodes from candidate list of nodes will be chosen at random by the staking contract to be part of the next epoch.
```

Account staking info:
	ID: 						 "e737ec6efbd26ef43bf676911cdc5a11ba15fc6562d05413e6589fccdd6c06d5"
	Initial Weight: 				 100
	Networking Address: 				 "flowaccess.mycompany.com:3569"
	Networking Key: 				 "f493a74704f6961ae7903e062ecd58d990672858eff99aece7bfbccf3aa02de8f1a624ecbf21a01e8b2f4a5854c231fbe218edd7762a34fea881f3958a215305"
	Role: 						 5
	Staking Key: 					 "ae8dcf81f3a70d72036b7ba2c586ed37ed0eb82b9c0a4aab998a8420f98894f94c14f84fa716e93654d3940fc0c8ff4d19b504c90a5b4918b28f421e9d3659dc2b7e246025ebeffea0d83cceefe315d7ed346dbe412fdac51b64997d97d29f7e"
	Tokens Committed: 				 0.00000000
	Tokens To Unstake: 				 100.00000000
	Tokens Rewarded: 				 0.00000000
	Tokens Staked: 					 100.00000000
	Tokens Unstaked: 				 0.00000000
	Tokens Unstaking: 				 0.00000000
	Node Total Stake (including delegators): 	 0.00000000
````

### Epoch setup event

Alternatively, if you can monitor events, look for [the epoch setup event](/staking/epoch-scripts-events/#epochsetup) that gets emitted by the epoch contract. That event is emitted at the end of epoch N's staking auction and contains a list of node IDs that are confirmed for the next epoch.

## Step 4 - Start your node

> If your node has gotten selected as part of Step 3, you can now start your node.

Reviewer 4: ```suggestion
If your node was selected as part of Step 3, you can now start your node.
```
